src/extension/ChronoController.hpp

```diff
@@ -74,65 +97,81 @@ namespace extension {
-                    if (NUClear::clock::now() + wait_offset > tasks.front().time) {
+                    // Acquire the mutex lock so we can wait on it
+                    std::unique_lock<std::mutex> lock(mutex);
 
-                        // Spinlock!
-                        while (NUClear::clock::now() < tasks.front().time) {
+                    // If we have no chrono tasks wait until we are notified
+                    if (tasks.empty()) {
+                        wait.wait(lock);
+                    }
+                    else {
+                        auto start  = NUClear::clock::now();
+                        auto target = tasks.front().time;
+
+                        if (target - start > cv_accuracy) {
+                            // Wait on the cv
+                            wait.wait_until(lock, target - cv_accuracy);
+
+                            // Update the accuracy of our cv wait
+                            const auto end   = NUClear::clock::now();
+                            const auto error = end - (target - cv_accuracy);  // when ended - when wanted to end
+                            if (error.count() > 0) {                          // only if we were late
+                                cv_accuracy = error > cv_accuracy ? error : ((cv_accuracy * 99 + error) / 100);
+                            }
                         }
+                        else if (target - start > ns_accuracy) {
+                            // Wait on nanosleep
+                            util::precise_sleep(target - start - ns_accuracy);
+
+                            // Update the accuracy of our precise sleep
+                            const auto end   = NUClear::clock::now();
+                            const auto error = end - (target - ns_accuracy);  // when ended - when wanted to end
+                            if (error.count() > 0) {                          // only if we were late
+                                ns_accuracy = error > ns_accuracy ? error : ((ns_accuracy * 99 + error) / 100);
+                            }
+                        }
+                        else {
+                            // Spinlock until we get to the time
+                            while (NUClear::clock::now() < tasks.front().time) {
+                            }
 
-                        const NUClear::clock::time_point now = NUClear::clock::now();
-
-                        // Move back from the end poping the heap
-                        for (auto end = tasks.end(); end != tasks.begin() && tasks.front().time < now;) {
                             // Run our task and if it returns false remove it
                             const bool renew = tasks.front()();
 
                             // Move this to the back of the list
-                            std::pop_heap(tasks.begin(), end, std::greater<>());
+                            std::pop_heap(tasks.begin(), tasks.end(), std::greater<>());
```

Comment:
If `nanosleep` or `cv.wait` had an error should we not also run the task?

Reply:
If they did then it would have bailed out of the if/else if and come back in.
If we select to sleep using one of the methods once it finishes we reenter